### ISSUE TITLE
Allow one ad to have a higher priority?

### ISSUE THREAD

**Issue by jpettitt** [ampproject/amphtml#4111]
We'd like to be able to designate one ad on a page as a higher load priority.  The current model of having all ads be priority 2 is creating a bad user experience.

Typically our publishers have a 320x50 ad at the top of the page.  On a page where the main content renders in ~600ms the ad doesn't fill until around 3.2 seconds.  Part of that is the nature of waterfall and ad exchanges but a good chunk of it is AMP just not calling the ad (it makes the call around 2300ms)

The net result of this is a page that changes visually after the initial load.

![image](https://cloud.githubusercontent.com/assets/524856/16924356/bb434540-4cd3-11e6-9381-b11d5f483103.png)

**Comment by jpettitt** [comment]
Example page if you want to observe it https://wesh.relaymedia.com/amp/national-news/how-a-freedom-fighter-became-a-lion-king-star/40416072

**Comment by lannka** [comment]
> The net result of this is a page that changes visually after the initial load.

@jpettitt I suppose you were not talking about a position jump after the ads loaded. It's just loaded very late, right?

**Comment by rudygalfi** [comment]
cc @jasti

**Comment by jpettitt** [comment]
@lannka yes just a really late load.

**Comment by jasti** [comment]
@jpettitt We'll definitely do early requests for A4A ads #3133 and considering for all ads. Please stay tuned.

**Comment by cramforce** [comment]
A slightly longer answer than you might have expected 😄 
https://medium.com/@cramforce/but-what-about-the-ads-bfe5025b6606#.uymgdsph5

We do not currently have plans to optimize ads beyond this.

**Comment by jpettitt** [comment]
@cramforce I love A4A, however we don't have any expectation that A4A creative will be around in volume any time soon.  In the mean time all of our pubs have a slow, above the fold, 320x50 ad unit.  you can see the problem on other sites too: nbcnews.com, independent.co.uk, businessinsider.com, telegraph.co.uk all have the same issue.  Allowing us to request that one ad be prioritized still won't be as fast as A4A but will  give a better UX.

**Comment by cramforce** [comment]
A4A will actually also make the legacy case faster due to cacheable ad
requests (renders at the same time as before, but the request is made
earlier and likely already done) and all simple doubleclick ads (just
images or text) will be automatically migrated.

**Comment by jasti** [comment]
@jpettitt If you are using DoubleClick, all ads will be requested early (earlier than they are today) when DFP's A4A implementation (https://github.com/ampproject/amphtml/blob/master/extensions/amp-ad-network-doubleclick-impl/0.1/doubleclick-a4a-config.js) goes live, which is expected by end of Aug. Essentially, getting what you asked for as part of this request.  
@michaelkleber to correct me, if I am wrong.

**Comment by michaelkleber** [comment]
As @cramforce and @jasti say, the A4A work will lead to ad _requests_ being issued much sooner, no matter what kind of ad gets returned.  But @jpettitt is quite right that, unless the response is an AMP creative, there will still be a deliberate delay before the ad gets _rendered_.

I don't think that's going to change.  Once a classical HTML+JS ad gets rendered, the AMP runtime loses the ability to make good on the performance guarantees it's making to the user.  So your attempt to improve UX by showing the ad earlier would actually degrade UX by making the rest of the page's interaction worse, and that is not a trade-off AMP is willing to make.

Of course AMP creatives will let us render the ad quickly _and_ honor performance guarantees, and this sort of highly-visible above-the-fold slot seems like a great place for publishers to try them out once #3133 launches.

**Comment by jasti** [comment]
Clarifying @michaelkleber's point further -  because the request is being sent out early (and therefore introducing parallelization of things like the ad auction) - in relative terms, even for a non-A4A response (accounting for deliberate delays in both cases), the ad will render faster than present day.

**Comment by jasti** [comment]
Once A4A for DoubleClick and Adsense is complete (expected this quarter), this will be less of an issue even for non-AMP format ads. Closing.
